File: BeFit iOS/ViewControllers/DeviceDetailViewController.m
//
//  DeviceDetailViewControllerUIViewController.m
//  BeFit iOS
//
//  Created by Jon on 2/6/17.
//  Copyright © 2017 Jon Brown. All rights reserved.
//

#import "DeviceDetailViewController.h"
#import "AppDelegate.h"

@interface DeviceDetailViewController ()

@end

@implementation DeviceDetailViewController
@synthesize device;


- (NSManagedObjectContext *)managedObjectContext {
    NSManagedObjectContext *context = nil;
    id delegate = [[UIApplication sharedApplication] delegate];
    if ([delegate performSelector:@selector(managedObjectContext)]) {
        context = [delegate managedObjectContext];
    }
    return context;
}


- (IBAction)cancel:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)save:(id)sender {
    
    if ([self.nameTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]].length == 0)
    {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:@"Please enter food list name" preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:nil];
        [alertController addAction:ok];
        
        [self presentViewController:alertController animated:YES completion:nil];
        
//        UIAlertView* alert = [[UIAlertView alloc] initWithTitle:@"Error" message:@"Please enter food list name" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil] ;
//        [alert show];
        return ;
    }
    
    if ([[AppDelegate CheckForDuplicateFoodList:_nameTextField.text] count] > 0)
    {
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:@"Food List already exists" preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"Okay" style:UIAlertActionStyleDefault handler:nil];
        [alertController addAction:ok];
        
        [self presentViewController:alertController animated:YES completion:nil];
        
       // [[[UIAlertView alloc] initWithTitle:@"Error" message:@"Food List already exists" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil]show];
        return ;
    }
    
    NSInteger maxID = [self GetMaxTranslationMaxId] + 1 ;
    NSManagedObjectContext *context = [self managedObjectContext];
    
    if (self.device) {
        // Update existing device
        [self.device setValue:self.nameTextField.text forKey:@"name"];
        [self.device setValue:[NSNumber numberWithInteger:maxID] forKey:@"orderIndex"];
        
    } else {
        // Create a new managed object
        FoodList* list = [NSEntityDescription insertNewObjectForEntityForName:@"FoodListObject" inManagedObjectContext:context];
        list.name =self.nameTextField.text;
        list.orderIndex = [NSNumber numberWithInteger:maxID] ;
        list.foods = [NSSet new];
//        [newDevice setValue:self.nameTextField.text forKey:@"name"];
//        [newDevice setValue:[NSNumber numberWithInteger:maxID] forKey:@"orderIndex"];
        
    }
    
    NSError *error = nil;
    // Save the object to persistent store
    if (![context save:&error]) {
        NSLog(@"Can't Save! %@ %@", error, [error localizedDescription]);
    }
    
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationbar.delegate = self;
    
    if (self.device) {
        [self.nameTextField setText:[self.device valueForKey:@"name"]];
    }
    
    // Do any additional setup after loading the view.
}

- (UIBarPosition)positionForBar:(id<UIBarPositioning>)bar {
    return UIBarPositionTopAttached;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}
-(int)GetMaxTranslationMaxId
{
    id appDelegate =(id)[[UIApplication sharedApplication] delegate];
    
    NSManagedObjectContext *context =[appDelegate managedObjectContext];
    
    NSEntityDescription *entityDesc =[NSEntityDescription entityForName:@"FoodListObject" inManagedObjectContext:context];
    
    NSFetchRequest *request = [[NSFetchRequest alloc] init];
    [request setEntity:entityDesc];
    
    NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"orderIndex" ascending:NO];
    NSArray *sortDescriptors = [[NSArray alloc] initWithObjects:sortDescriptor, nil];
    [request setSortDescriptors:sortDescriptors];
    
    
    [request setFetchLimit:1];
    
    NSError *error = nil;
    NSArray *results = [[appDelegate managedObjectContext] executeFetchRequest:request error:&error];
    
    int maximumValue = 0;
    if (results.count == 1)
    {
        FoodList *result = (FoodList*)[results objectAtIndex:0];
        maximumValue =  [result.orderIndex intValue];
        NSLog(@"The max value is for request is :%d",maximumValue );
    }
    return  maximumValue;
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: BeFit iOS/ViewControllers/DeviceDetailViewController.swift
//
//  DeviceDetailViewController.swift
//  BeFitTracker
//
//  Created by Andrej Slegl on 10/25/19.
//  Copyright © 2019 Jon Brown. All rights reserved.
//

import Foundation
import UIKit
import CoreData

protocol DeviceDetailViewControllerDelegate: AnyObject {
    func deviceDetailViewControllerDidSave()
}

class DeviceDetailViewController: UIViewController,
UINavigationBarDelegate {
    //MARK: Properties
    @IBOutlet private weak var nameTextField: UITextField!
    
    weak var delegate: DeviceDetailViewControllerDelegate?
    var device: FoodList?
    
    private var managedObjectContext: NSManagedObjectContext? {
        return (UIApplication.shared.delegate as? AppDelegate)?.managedObjectContext
    }
    
    //MARK: Init
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        if let device = device {
            nameTextField.text = device.name
        }
    }
    
    //MARK: - Actions
    
    @IBAction private func cancel(_ sender: AnyObject) {
        dismiss(animated: true, completion: nil)
    }
    
    @IBAction private func save(_ sender: AnyObject) {
        if nameTextField.text?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true {
            let alertController = UIAlertController(title: "Error", message: "Please enter food list name", preferredStyle: .alert)
            let ok = UIAlertAction(title: "Okay", style: .default, handler: nil)
            alertController.addAction(ok)
            
            present(alertController, animated: true)
            
    //        UIAlertView* alert = [[UIAlertView alloc] initWithTitle:@"Error" message:@"Please enter food list name" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil] ;
    //        [alert show];
            return
        }
        
        if !AppDelegate.check(forDuplicateFoodList: nameTextField.text ?? "").isEmpty {
            let alertController = UIAlertController(title: "Error", message: "Food List already exists", preferredStyle: .alert)
            let ok = UIAlertAction(title: "Okay", style: .default, handler: nil)
            alertController.addAction(ok)
            
            present(alertController, animated: true)
            
           // [[[UIAlertView alloc] initWithTitle:@"Error" message:@"Food List already exists" delegate:nil cancelButtonTitle:@"Okay" otherButtonTitles: nil]show];
            return
        }
        
        let maxID = getMaxTranslationMaxId() + 1
        
        guard let context = managedObjectContext else {
            assertionFailure()
            return
        }
        
        if let device = device {
            // Update existing device
            device.name = nameTextField.text ?? ""
            device.orderIndex = NSNumber(value: maxID)
        } else {
            let list = NSEntityDescription.insertNewObject(forEntityName: "FoodListObject", into: context) as! FoodList
            list.name = nameTextField.text ?? ""
            list.orderIndex = NSNumber(value: maxID)
            list.foods = NSSet()
        }
        
        do {
            try context.save()
        } catch let error {
            print("Can't Save!", error.localizedDescription)
        }
        
        delegate?.deviceDetailViewControllerDidSave()
        dismiss(animated: true, completion: nil)
    }
    
    //MARK: - Private
    
    private func getMaxTranslationMaxId() -> Int {
        guard let context = managedObjectContext else {
            assertionFailure()
            return 0
        }
        
        let request = FoodList.fetchRequest() as NSFetchRequest<FoodList>
        let sortDescriptor = NSSortDescriptor(key: "orderIndex", ascending: false)
        request.sortDescriptors = [sortDescriptor]
        request.fetchLimit = 1
        
        do {
            let results = try context.fetch(request)
            
            guard let firstResult = results.first else {
                return 0
            }
            
            let maximumValue = firstResult.orderIndex?.intValue ?? 0
            print("The max value is for request is:", maximumValue)
            
            return maximumValue
        } catch let error {
            print("Failed to fetch:", error.localizedDescription)
            return 0
        }
    }
}
